src/ai/backend/manager/services/agent/service.py

```diff
@@ -141,3 +172,55 @@ async def recalculate_usage(
     ) -> RecalculateUsageActionResult:
         await self._agent_registry.recalc_resource_usage()
         return RecalculateUsageActionResult()
+
+    async def handle_heartbeat(self, action: HandleHeartbeatAction) -> HandleHeartbeatActionResult:
+        now = datetime.now(tzutc())
+        reported_agent_info = action.agent_info
+
+        reported_agent_state_sync_data = AgentStateSyncData(
+            now=now,
+            slot_key_and_units={
+                SlotName(k): SlotTypes(v[0])
+                for k, v in reported_agent_info["resource_slots"].items()
+            },
+            current_addr=reported_agent_info["addr"],
+            public_key=reported_agent_info["public_key"],
+        )
+
+        async with self._heartbeat_lock:
+            upsert_data = AgentHeartbeatUpsert.from_agent_info(
+                agent_id=action.agent_id,
+                agent_info=action.agent_info,
+                heartbeat_received=now,
+            )
+            result: UpsertResult = await self._agent_repository.sync_agent_heartbeat(
+                action.agent_id,
+                upsert_data,
+                reported_agent_state_sync_data,
+            )
+            if result.need_agent_cache_update:
```

Comment:
The field "need agent cache update" is excessive; please return only whether it is newly registered or whether the value has been updated.